layers/dir.go

```diff
@@ -0,0 +1,63 @@
+package layers
+
+import (
+	"archive/tar"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+
+	"github.com/buildpacks/lifecycle/archive"
+)
+
+// DirLayer creates a layer form the given directory
+func (f *Factory) DirLayer(id string, dir string) (layer Layer, err error) {
+	dir, err = filepath.Abs(dir)
+	if err != nil {
+		return Layer{}, err
+	}
+
+	tarPath := filepath.Join(f.ArtifactsDir, escape(id)+".tar")
+	if f.tarHashes == nil {
+		f.tarHashes = make(map[string]string)
+	}
+	if sha, ok := f.tarHashes[tarPath]; ok {
+		f.Logger.Debugf("Reusing tarball for layer %q with SHA: %s\n", id, sha)
+		return Layer{
+			ID:      id,
+			TarPath: tarPath,
+			Digest:  sha,
+		}, nil
+	}
+	lw, err := newFileLayerWriter(tarPath)
+	if err != nil {
+		return Layer{}, err
+	}
+	defer func() {
+		if closeErr := lw.Close(); err == nil {
+			err = closeErr
+		}
+	}()
+	tw := archive.NewNormalizingTarWriter(tar.NewWriter(lw))
+	parents, err := parents(dir)
+	if err != nil {
+		return Layer{}, err
+	}
+	if err := archive.WriteFilesToArchive(tw, parents); err != nil {
+		return Layer{}, err
+	}
+	tw.WithUID(f.UID)
```

Comment:
Can/Should the logic of writing parent directories without `uid`/`gid` be encapsulated in `NormalizingTarWriter`? I can see that `addParentDirs` had the same behavior.

Reply:
I wanted to keep the rules like "don't change perms for dir parents" in the `layers` package so that the `NomalizingTarWriter` and the entire `archive` package can just include generic utilities. Encapsulating this in `NormalizingTarWriter` would require the `NormalizingTarWriter` to be context aware (it needs to know whether a header is a parent dir) and make it less simple and less reusable across contents.